File: server/Api/PublicRoutes.java
package server.Api;

import java.sql.Connection;
import java.sql.SQLException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import server.Commands.ContactOptOutCommand;
import server.Commands.LinkClickedCommand;
import server.Data.DB;
import server.Models.Contact;
import server.Models.SignUpRequest;

@RestController
public class PublicRoutes {

    @Autowired
    private DB db;

    @GetMapping("/about")
    public ResponseEntity<String> about() {
        return new ResponseEntity<>("Tailwind Traders Mail Services API", HttpStatus.OK);
    }

    @GetMapping("/unsubscribe/{key}")
    public ResponseEntity<Boolean> unsubscribe(@PathVariable String key) {
        try (Connection conn = db.connect()) {
            ContactOptOutCommand cmd = new ContactOptOutCommand(key);
            boolean result = cmd.execute(conn).getUpdated() > 0;
            return new ResponseEntity<>(result, HttpStatus.OK);
        } catch (SQLException e) {
            return new ResponseEntity<>(false, HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    @GetMapping("/link/clicked/{key}")
    public ResponseEntity<String> linkClicked(@PathVariable String key) {
        LinkClickedCommand cmd = new LinkClickedCommand(key);
        String result = cmd.execute();
        return new ResponseEntity<>(result, HttpStatus.OK);
    }

    @PostMapping("/signup")
    public ResponseEntity<Integer> signUp(@RequestBody SignUpRequest req) {
        Contact contact = new Contact(req.getName(), req.getEmail());
        try (Connection conn = db.connect()) {
            int result = conn.createQuery("INSERT INTO contacts (email, name) VALUES (?, ?)")
                             .addParameter(contact.getEmail())
                             .addParameter(contact.getName())
                             .executeUpdate();
            return new ResponseEntity<>(result, HttpStatus.OK);
        } catch (SQLException e) {
            return new ResponseEntity<>(0, HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
}

File: server/Api/PublicRoutes.cs
//public endpoints for subscribe/unsubscribe
using System.Data;
using Dapper;
using Microsoft.AspNetCore.Mvc;
using Tailwind.Data;
using Tailwind.Mail.Commands;
using Tailwind.Mail.Models;

namespace Tailwind.Mail.Api;

public class PublicRoutes{

  public static void MapRoutes(IEndpointRouteBuilder app)
  {

    //public routes
    app.MapGet("/about", () => "Tailwind Traders Mail Services API").WithOpenApi(op => {
      op.Summary = "📋 Get API Information";
      op.Description = "Returns basic information about the Tailwind Traders Mail Services API, including version and capabilities.";
      return op;
    }).WithTags("Public");

    app.MapGet("/unsubscribe/{key}", (string key, [FromServices] IDb db) => {
      using(var conn = db.Connect()){
        var cmd = new ContactOptOutCommand(key);
        var result = cmd.Execute(conn);
        return result.Updated > 0;
      }
    }).WithOpenApi(op => {
      op.Summary = "🚫 Unsubscribe Contact";
      op.Description = "Allows a contact to opt-out from the mailing list using their unique unsubscribe key. This key is typically included in email footers.";
      op.Parameters[0].Description = "The contact's unique unsubscribe key (provided in email communications)";
      return op;
    }).WithTags("Public");

    //this isn't implemented yet in terms of data
    app.MapGet("/link/clicked/{key}", (string key, [FromServices] IDb db) => {
      var cmd = new LinkClickedCommand(key);
      var result = cmd.Execute();
      return result;
    }).WithOpenApi(op => {
      op.Summary = "🔗 Track Link Click";
      op.Description = "Records when a contact clicks a tracked link in an email. This endpoint collects engagement metrics for broadcasts and email sequences.";
      op.Parameters[0].Description = "The unique tracking key for the clicked link (embedded in email links)";
      return op;
    }).WithTags("Public");

    app.MapPost("/signup", async ([FromBody] SignUpRequest req,  [FromServices] IDb db) => {
      var contact = new Contact{
        Email = req.Email,
        Name = req.Name
      };
      using var conn = db.Connect();
      var result = await conn.ExecuteAsync("insert into contacts (email, name) values (@Email, @Name)", contact);
      return result;
      
    }).WithOpenApi(op => {
      op.Summary = "✉️ Sign Up New Contact";
      op.Description = "Adds a new contact to the mailing list. This endpoint is typically used by website signup forms to collect new subscribers.";
      op.RequestBody.Description = "Contact information including name and email address";
      return op;
    }).WithTags("Public");
  }

}